sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitionsSuite.scala

```diff
@@ -140,6 +140,24 @@ class PruneFileSourcePartitionsSuite extends PrunePartitionSuiteBase with Shared
     }
   }
 
+  test("SPARK-40565: don't push down non-deterministic filters for V2 file sources") {
```

Comment:
BTW, this query seems to fail in V1 source. Could you confirm this?
```scala
-    withSQLConf((SQLConf.USE_V1_SOURCE_LIST.key, "")) {
+    withSQLConf((SQLConf.USE_V1_SOURCE_LIST.key, "parquet")) {
```

Reply:
So it looks like that's because of the way the "number of partitions" is calculated for V2 reads at https://github.com/apache/spark/blob/697574c2d5014626d2c960b222b41a8da7c702bf/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/PruneFileSourcePartitionsSuite.scala#L168. It uses `BatchScanExec.partitions.size`, which is the number of RDD partitions, _not_ file partitions. I can see if I can fix that but don't see an easy way like `FileSourceScanExec.selectedPartitions`, there's no intermediate variable to get from a batch scan. Might just have to try to calculate unique partitions values from all files to get actual number of file partitions?

If you change the 2 to 3 for the tests it passes for V1.